### PR TITLE
community: Add mmr and similarity_score_threshold retrieval to DatabricksVectorSearch

### PR DESCRIPTION
**Pull request by david-tempelmann**
- **Description:** This PR adds support for `search_types="mmr"` and `search_type="similarity_score_threshold"` to retrievers using `DatabricksVectorSearch`, 
  - **Issue:** 
  - **Dependencies:**
  - **Twitter handle:**

### CODE REVIEW

File: libs/community/langchain_community/vectorstores/databricks_vector_search.py [hunk 1]
```diff
@@ -321,6 +324,134 @@ def similarity_search_with_score(
...
+
+        mmr_selected = maximal_marginal_relevance(
+            np.array(embedding, dtype=np.float32),
+            embeddings,
+            k=k,
+            lambda_mult=lambda_mult,
+        )
+
+        candidates = self._parse_search_response(
+            search_resp, ignore_cols=[embedding_column]
```
Comment (freemso) [inline libs/community/langchain_community/vectorstores/databricks_vector_search.py:450]: An edge case: what if the `self.columns` contains the embedding column, meaning the customer *explicitly* asking to keep embeddings in the metadata of the returned Documents?

File: libs/community/tests/unit_tests/vectorstores/test_databricks_vector_search.py [hunk 2]
```diff
@@ -456,6 +482,24 @@ def test_similarity_search(index_details: dict) -> None:
...
+    query = "foo"
+    filters = {"some filter": True}
+    limit = 1
+
+    search_result = vectorsearch.max_marginal_relevance_search(
+        query, k=limit, filters=filters
+    )
+    assert [doc.page_content for doc in search_result] == ["foo"]
+
+
```
Comment (freemso) [inline libs/community/tests/unit_tests/vectorstores/test_databricks_vector_search.py:502]: Could you also add some unit tests to test if the below code works:
```
# Retrieve more documents with higher diversity
# Useful if your dataset has many similar documents
vectorsearch.as_retriever(
    search_type="mmr",
    search_kwargs={'k': 6, 'lambda_mult': 0.25}
)

# Fetch more documents for the MMR algorithm to consider
# But only return the top 5
vectorsearch.as_retriever(
    search_type="mmr",
    search_kwargs={'k': 5, 'fetch_k': 50}
)

# Only retrieve documents that have a relevance score
# Above a certain threshold
vectorsearch.as_retriever(
    search_type="similarity_score_threshold",
    search_kwargs={'score_threshold': 0.8}
)
```

File: libs/community/langchain_community/vectorstores/databricks_vector_search.py [hunk 3]
```diff
@@ -321,6 +324,134 @@ def similarity_search_with_score(
...
+        """
+
+        def _db_relevance_score_fn(distance: float) -> float:
+            """
+            Scores returned by the Databricks Vector Search API are
+            normalized cosine similarity scores on a scale [0, 1] -
+            1 means most similar
+            """
+
+            return distance
```
Comment (freemso) [inline libs/community/langchain_community/vectorstores/databricks_vector_search.py:344]: nit: Instead of having this nested function, can we convert this to a class-level static method similar to [here](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/vectorstores/elasticsearch.py#L699)?

File: libs/community/langchain_community/vectorstores/databricks_vector_search.py [hunk 4]
```diff
@@ -321,6 +324,134 @@ def similarity_search_with_score(
...
+        - the scale of your embeddings (OpenAI's are unit normed. Many others are not!)
+        - embedding dimensionality
+        - etc.
+        """
+
+        def _db_relevance_score_fn(distance: float) -> float:
+            """
+            Scores returned by the Databricks Vector Search API are
+            normalized cosine similarity scores on a scale [0, 1] -
+            1 means most similar
```
Comment (freemso) [inline libs/community/langchain_community/vectorstores/databricks_vector_search.py:341]: Can we confirm the similarity metric used by Databricks Vector Search?

**Review by freemso — COMMENTED** [review]
Thanks for the changes! Left some comments. Mostly looks good. But please get confirmation of the similarity metric from Databricks Vector Search team.

File: libs/community/langchain_community/vectorstores/databricks_vector_search.py (hunk 1 above)
Comment (david-tempelmann) [reply libs/community/langchain_community/vectorstores/databricks_vector_search.py]: Should be covered now.

File: libs/community/tests/unit_tests/vectorstores/test_databricks_vector_search.py (hunk 2 above)
Comment (david-tempelmann) [reply libs/community/tests/unit_tests/vectorstores/test_databricks_vector_search.py:595]: Done. Please have a look.

File: libs/community/langchain_community/vectorstores/databricks_vector_search.py (hunk 3 above)
Comment (david-tempelmann) [reply libs/community/langchain_community/vectorstores/databricks_vector_search.py]: Makes sense. I've changed it.

File: libs/community/langchain_community/vectorstores/databricks_vector_search.py (hunk 4 above)
Comment (david-tempelmann) [reply libs/community/langchain_community/vectorstores/databricks_vector_search.py]: This has now been confirmed to be a normalized score 1/(1+d) with d being the L2-distance. I've added that information to the code.

**Review by freemso — APPROVED** [review]
LGTM! Thanks for iterating on this!

**Comment by david-tempelmann** [comment]
@freemso Sorry, I just realized that there was an issue with filters in the mmr implementation and that there was no test for it so I added it. Now everything should be ok.

**Review by freemso — APPROVED** [review]

**Review by baskaryan — APPROVED** [review]
